[v4] Composition engine (#404)
V4Fire/Core: src/core/request/engines/composition/README.md

  {
    request: (opts, params, {boundRequest}) => boundRequest(request('http://locahost:5555/json/1')),
    as: 'val1',
    failCompositionOnError: true
  },
  {
    request: (opts, params, {boundRequest}) => boundRequest(request('http://locahost:5555/json/2')),
    as: 'val2'
  },
])});

try {
  await r('').data;
} catch (error) {
  // If the first request encountered an error, the engine will not return any data,
  // it will crash with an error and return the error with which the first request ended.
}
```

If both request objects have the failCompositionOnError flag set, the error will be thrown as soon as any of them responds with an error.
However, this behavior can be changed by setting the aggregateErrors flag,
which will alter the engine's behavior in this situation.
If the flag is set to true, the engine will wait for all requests to complete and gather errors from
all requests that have the failCompositionOnError flag set.
Additionally, instead of the request's own error,
the engine will return an AggregateError that will contain all the errors from the requests.

```typescript
const r = request({engine: compositionEngine([
  {

>     request: (opts, params, {boundRequest}) => boundRequest(request('http://locahost:5555/json/1')),

Reviewer: А нельзя избавиться от потребности явно вызывать bound?

    as: 'val2',
    failCompositionOnError: true
  },
], aggregateErrors: true)});

try {
  await r('').data;
} catch (error) {
  // If one or both requests end with errors, the error here will be an AggregateError
  // that will contain the errors from the requests.
}
```

### Using as a Provider Request Engine

#### Basic Usage

Let's say we want to have a provider that makes 3 requests, two in parallel and one depending on the response of the first request.
Let's implement a provider that works this way:

```typescript
@provider
class Provider1 extends Provider {
  override baseURL: string = 'http://localhost:5000/json/1';
}

@provider
class Provider2 extends Provider {
  override baseURL: string = 'http://localhost:5000/json/2';
}

@provider
class Provider3 extends Provider {
  override baseURL: string = 'http://localhost:5000/json/3';
}

@provider
class MainProvider extends Provider {
  static override request: Provider['request'] = Provider.request({
    engine: compositionEngine([
      {
        request: (opts, params, {boundRequest}) => {
          const result = {
            val1: undefined,
            val2: undefined

          }

>           const val1 = await boundRequest(new Provider1()).get().data;

Reviewer: Вот тут кейс валидный

            return {
              getData: data
            }
          }

          const data = await boundRequest(new Provider1()).post().data;

          return {
            postData: data
          }
        },
        as: 'request1'
      },
      {
        request: () => new Provider2().get(),
        as: 'request2'
      }
    ])
  });
}

const getData = await new MainProvider().get().data; // {request1: {getData: data}, request2: data};
const postData = await new MainProvider().post().data // {request1: {postData: data}, request2: data};
```

#### Spread Data Into Result

There is also an option to merge the result of a request with the resulting object
instead of storing it in a specific property of the resulting object.
To do this, specify 'spread' in the 'as' field, and the result of
the request will be merged with the resulting object.

```typescript
export class MyCompositionProvider extends Provider {
  static override request: typeof Provider.request = Provider.request({
    engine: compositionEngine([
      {
        request: () => new Provider1().get() // returns {foo: 'value', bar: 'value'}

>         as: 'spread'

Reviewer 2: не очень хорошо использовать нейминг, который потенциально может конфликтить с валидным именем, можно заменить на:
- '[[SPREAD]]'
- '...'
- или вообще вместо `as` использовать булево значение `spead: true`

      - [Spread Data Into Result](#spread-data-into-result)
      - [Passing constructor parameters of a provider to child providers](#passing-constructor-parameters-of-a-provider-to-child-providers)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# core/request/engines/composition

This module provides an API (engine) for creating a composition of requests.
This engine can be used as an engine for request and, accordingly, as an engine for providers.
Mostly, this engine is designed specifically for use with providers.

## Usage

### Creating a Request Composition

Let's imagine that you frequently need to make requests to several different APIs.
To solve this problem, you can use a query composition engine:

```typescript
import request from 'core/request';
import { compositionEngine } from 'core/request/engines/composition';

const r = request({engine: compositionEngine([
  {
    request: () => request('http://locahost:5555/json/1'),
    as: 'val1'
  },
  {
    request: () => request('http://locahost:5555/json/2'),
    as: 'val2'
  },
])});

const data = await r('').data; // {val1: data, val2: data}
```

> It is possible to use other approaches, for example, creating a simple function that will make several requests, and there is no arguing with you.

Reviewer 2: Смущает фраза:  `and there is no arguing with you.`

But as mentioned in the module description, first and foremost, this engine is needed to expand the capabilities of data providers.
Let's now consider how this engine helps to enhance the abilities of data providers.

### Using the `requestFilter` Property

> Each «request» object has the `requestFilter` property.

Reviewer 2: `«request»` такие кавычки в английском языке не используются

The `requestFilter` is called before the request function is called.
If the `requestFilter` returns false, the request will not be called and, consequently,
there will be no data from this request in the response.
This allows you to implement various scenarios in which you, depending,
for example, on the query parameters, need to make one or another request.

```typescript
const r = request({engine: compositionEngine([
  {
    request: () => request('http://locahost:5555/json/1'),
    as: 'val1',
    requestFilter: (opts, params) => opts.query.auth != null
  },
  {
    request: () => request('http://locahost:5555/json/2'),
    as: 'val2'
  },
])});
```

### Handling Request Errors

By default, if an error occurs in the `request` function, the engine will ignore it and continue making other requests.
However, this behavior can be changed. For example, you can set the `failCompositionOnError` property of the request object to `true`.
In this case, if there is an error in the request function of this request object,
it will be thrown as a `ComposedProvider` request error.

```typescript
const r = request({engine: compositionEngine([
  {

>     request: (opts, params, {boundRequest}) => boundRequest(request('http://locahost:5555/json/1')),

Reviewer 2: А что за boundRequest? Непонятно, нужно пояснение

    as: 'val1',
    failCompositionOnError: true
  },
  {
    request: (opts, params, {boundRequest}) => boundRequest(request('http://locahost:5555/json/2')),
    as: 'val2'
  },
])});

try {
  await r('').data;
} catch (error) {
  // If the first request encountered an error, the engine will not return any data,
  // it will crash with an error and return the error with which the first request ended.
}
```

If both request objects have the `failCompositionOnError` flag set, the error will be thrown as soon as any of them responds with an error.
However, this behavior can be changed by setting the aggregateErrors flag,
which will alter the engine's behavior in this situation.

> If the flag is set to true, the engine will wait for all requests to complete and gather errors from

Reviewer 2: ```suggestion
If the flag is set to true, the engine will wait for all requests that have the `failCompositionOnError` flag to complete.
```

Reviewer 2: Это можно дропнуть `all requests that have the `failCompositionOnError` flag set.`

all requests that have the `failCompositionOnError` flag set.

> Additionally, instead of the request's own error,

Reviewer 2: ```suggestion
In this case the engine will return an AggregateError that will contain all the errors from the requests.
```

  override baseURL: string = 'http://localhost:5000/json/3';
}

@provider
class MainProvider extends Provider {
  static override request: Provider['request'] = Provider.request({
    engine: compositionEngine([
      {
        request: ({boundRequest}) => {
          const result = {
            val1: undefined,
            val2: undefined

          }

          const val1 = await boundRequest(new Provider1()).get().data;
          result.val1 = val1;

          if (val1.needAnotherRequest) {
            const val2 = await boundRequest(new Provider2()).get().data;
            result.val2 = val2;
          }

          return result;
        },
        as: 'request1'
      },
      {
        request: () => new Provider3().get(),
        as: 'request2'
      }
    ])
  });
}

const data = await new MainProvider().get().data;
console.log(data) // {request1: {val1: data, val2: data}, request2: data}
```

Notice that, the request function provides a special function called `boundRequest`,
which is needed to bind the request object to the provider engine. The request object returned
from the request function is automatically wrapped in this function. However, if you make 

> multiple requests within the request function, it is important to use `boundRequest` to avoid memory leaks.

Reviewer 3: Кажется важное замечание, может как-то выделить это? Правда не знаю как лучше)

Reviewer 3: Хотя, возможно это и не поможет, пускай внимательнее читают доку))

Author: Ну если не читал доку то не поможет )
Не важно выделять или нет

```typescript
@provider
class Provider1 extends Provider {
  override baseURL: string = 'http://localhost:5000/json/1';
}

@provider
class Provider2 extends Provider {
  override baseURL: string = 'http://localhost:5000/json/2';
}

class MainProvider extends Provider {
  static override request: Provider['request'] = Provider.request({
    engine: compositionEngine([
      {
        request: (opts, params, {boundRequest}) => {
          if (opts.method === 'GET') {
            const data = await boundRequest(new Provider1()).get().data;

            return {
              getData: data
            }
          }

          const data = await boundRequest(new Provider1()).post().data;

          return {
            postData: data
          }
        },
        as: 'request1'
      },
      {
        request: () => new Provider2().get(),
        as: 'request2'
      }
    ])
  });
}

const getData = await new MainProvider().get().data; // {request1: {getData: data}, request2: data};
const postData = await new MainProvider().post().data // {request1: {postData: data}, request2: data};
```

> #### Spread Data Into Result

Reviewer 2: Мб, добавить ещё общий коллбэк, который получает на вход все ответы от провайдеров и позволяет их трансформировать как угодно? Например:

```js

 engine: providerCompositionEngine([
      {
        request: () => new Provider1().get() // returns {foo: 'value', bar: 'value'}
        as: 'spread'
      },
      {
        request: () => new Provider2().get() // returns {data: [...]}
        as: 'banners'
      }
    ],
   {
        transform: (result) => {
          const {foo, bar, banners} = result;
          // do something with foo, bar, banners
          return {...};
        }
    }
);

```

Author: Для этого есть декодеры

Author: Просто при миграции с основных провайдеров частый кейс что надо оставить все как есть (чтобы не переделывать db.field на db.content.field), сначала я подумал что можно декодером но постоянно писать эту мерж функцию (или импортить ее откуда-то)  сомнительное удовольствие, поэтому принял решение добавить такой функционал

There is also an option to merge the result of a request with the resulting object
instead of storing it in a specific property of the resulting object.
To do this, specify 'spread' in the 'as' field, and the result of
the request will be merged with the resulting object.

```typescript
export class MyCompositionProvider extends Provider {
  static override request: typeof Provider.request = Provider.request({

>     engine: providerCompositionEngine([

Reviewer 2: `providerCompositionEngine` - такого не видел в index.ts
